src/main/java/com/sap/sgs/phosphor/fosstars/data/github/GitHubDataFetcher.java

```diff
@@ -0,0 +1,171 @@
+package com.sap.sgs.phosphor.fosstars.data.github;
+
+import com.sap.sgs.phosphor.fosstars.data.GitHubDataCache;
+import com.sap.sgs.phosphor.fosstars.tool.github.GitHubProject;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.kohsuke.github.GHCommit;
+import org.kohsuke.github.GHRepository;
+import org.kohsuke.github.GitHub;
+
+/**
+ * Helper class for GitHub data providers which pulls the data from a GitHub repository.
+ */
+public class GitHubDataFetcher {
+
+  /**
+   * The singleton instance.
+   */
+  private static GitHubDataFetcher INSTANCE;
+
+  /**
+   * An interface to the GitHub API.
+   */
+  private final GitHub github;
+
+  /**
+   * A private constructor for singleton pattern.
+   */
+  private GitHubDataFetcher(GitHub github) {
+    this.github = github;
+  }
+
+  private final GitHubDataCache<List<GHCommit>> commitsCache =
+      new GitHubDataCache<List<GHCommit>>();
+
+  private final GitHubDataCache<GHRepository> repositoryCache = new GitHubDataCache<GHRepository>();
+
+  /**
+   * Creates a single instance of this helper class.
+   * 
+   * @return instance of this data pull helper.
+   */
+  public static GitHubDataFetcher instance(GitHub github) {
+    if (INSTANCE == null) {
+      INSTANCE = new GitHubDataFetcher(github);
+    }
+    return INSTANCE;
+  }
+  
+  /**
+   * Gets the list of commits for a GitHub project repository.
+   * 
+   * @param project of type {@link GitHubProject}, which holds the project information.
+   * @return List of {@link GHCommit}.
+   * @throws IOException occurred during REST call to GitHub API.
+   */
+  private List<GHCommit> commitsFor(GitHubProject project) throws IOException {
+    if (commitsCache.containsKey(project)) {
+      Optional<List<GHCommit>> commits = commitsCache.get(project);
+      if (commits.isPresent()) {
+        return commits.get();
+      }
+    }
```

Comment:
The `get()` method doesn't return null, does it? This could be a bit shorter and without nested if-blocks:

```java
Optional<List<GHCommit>> commits = commitsCache.get(project);
if (commits.isPresent()) {
  return commits.get();
}
```

Reply:
`isPresent()` should check for that. And anyways `commitsCache.get(project);` returns an `Optional.empty()` if the content is not there.

Reply:
Okay, let's make the code shorter then.